src/Analyzer/Resolve/QueryAnalyzer.cpp

```diff
@@ -1154,195 +1151,226 @@ std::string QueryAnalyzer::rewriteAggregateFunctionNameIfNeeded(
+            identifier_resolve_context.allow_to_check_join_tree = false;
 
-            bool scope_was_query = scope_to_check->scope_node->getNodeType() == QueryTreeNodeType::QUERY;
-            scope_to_check = scope_to_check->parent_scope;
+            /** From parent scopes we can resolve table identifiers only as CTE.
+                * Example: SELECT (SELECT 1 FROM a) FROM test_table AS a;
+                *
+                * During child scope table identifier resolve a, table node test_table with alias a from parent scope
+                * is invalid.
+                */
+            identifier_resolve_context.allow_to_check_aliases = false;
+        }
 
-            if (scope_was_query)
-                break;
+        if (!scope.context->getSettingsRef()[Setting::enable_global_with_statement])
+        {
+            identifier_resolve_context.allow_to_check_aliases = false;
+            identifier_resolve_context.allow_to_check_cte = false;
         }
     }
 
-    if (!scope.context->getSettingsRef()[Setting::enable_global_with_statement])
+    identifier_resolve_context.allow_to_check_database_catalog = false;
+
+    auto resolve_result = tryResolveIdentifier(identifier_lookup, *scope.parent_scope, identifier_resolve_context);
+    auto & resolved_identifier = resolve_result.resolved_identifier;
+
+    if (!resolved_identifier)
         return {};
 
-    /** Nested subqueries cannot access outer subqueries table expressions from JOIN tree because
-      * that can prevent resolution of table expression from CTE.
-      *
-      * Example: WITH a AS (SELECT number FROM numbers(1)), b AS (SELECT number FROM a) SELECT * FROM a as l, b as r;
-      */
+    /** From parent scopes we can resolve table identifiers only as CTE.
+        * Example: SELECT (SELECT 1 FROM a) FROM test_table AS a;
+        *
+        * During child scope table identifier resolve a, table node test_table with alias a from parent scope
+        * is invalid.
+        */
     if (identifier_lookup.isTableExpressionLookup())
-        identifier_resolve_settings.allow_to_check_join_tree = false;
-
-    while (scope_to_check != nullptr)
     {
-        auto lookup_result = tryResolveIdentifier(identifier_lookup, *scope_to_check, identifier_resolve_settings);
-        const auto & resolved_identifier = lookup_result.resolved_identifier;
+        auto * subquery_node = resolved_identifier->as<QueryNode>();
+        auto * union_node = resolved_identifier->as<UnionNode>();
 
-        scope_to_check = scope_to_check->parent_scope;
+        bool is_cte = (subquery_node && subquery_node->isCTE()) || (union_node && union_node->isCTE());
+        bool is_table_from_expression_arguments = resolve_result.isResolvedFromExpressionArguments() &&
+            resolved_identifier->getNodeType() == QueryTreeNodeType::TABLE;
+        bool is_from_join_tree_or_aliases = !initial_scope_is_query && (resolve_result.isResolvedFromJoinTree() || resolve_result.isResolvedFromAliases());
```

Comment:
I really need a comment for this.
Maybe, this is a duplicating logic.
Because previously we disabled
```
allow_to_check_join_tree = false;
allow_to_check_aliases = false;
```
but now we also re-check that the result is from the query tree or alias.

Reply:
Nope, it's not. `allow_to_check_join_tree` and `allow_to_check_aliases` are disabled if the initial scope is query. It's allowed to resolve identifier from parent aliases or join tree inside of lambda body.

```sql
SELECT map(x > x in a, [1, 2]) FROM numbers(3) as a;
```